File: typescript/src/prompts/generator.ts
import type { ToolRegistry } from "../tools/index.js";

export interface PromptGenerator {
  generateToolSection(registry: ToolRegistry): string;
  generateToolSchemas(registry: ToolRegistry): object[];
}

export class DefaultPromptGenerator implements PromptGenerator {
  generateToolSection(registry: ToolRegistry): string {
    const tools = registry.all();
    if (tools.length === 0) return "";

    const lines = ["## Available Tools\n"];
    for (const tool of tools) {
      lines.push(`- **${tool.name}**: ${tool.description}`);
    }
    return lines.join("\n");
  }

  generateToolSchemas(registry: ToolRegistry): object[] {
    return registry.all().map((tool) => ({
      type: "function",
      function: {
        name: tool.name,
        description: tool.description,
        parameters:
          "inputSchema" in tool
            ? (tool as unknown as { inputSchema(): Record<string, unknown> }).inputSchema()
            : {},
      },
    }));
  }
}

File: python/src/pedro_agentware/prompts/generator.py
"""Prompt generator for tools."""

from typing import Any, Protocol

from ..tools import ToolRegistry


class PromptGenerator(Protocol):
    """Protocol for prompt generators."""

    def generate_tool_section(self, registry: ToolRegistry) -> str:
        """Generate the formatted tool documentation block."""
        ...

    def generate_tool_schemas(self, registry: ToolRegistry) -> list[dict[str, Any]]:
        """Generate JSON schema block for native tool calling."""
        ...


class DefaultPromptGenerator:
    """Default implementation of PromptGenerator."""

    def __init__(self) -> None:
        pass

    def generate_tool_section(self, registry: ToolRegistry) -> str:
        """Generate the tool section."""
        tools = registry.all()
        if not tools:
            return ""

        lines = ["## Available Tools\n"]
        for tool in tools:
            lines.append(f"- **{tool.name}**: {tool.description}")
        return "\n".join(lines)

    def generate_tool_schemas(self, registry: ToolRegistry) -> list[dict[str, Any]]:
        """Generate JSON schemas for tools."""
        return [
            {
                "type": "function",
                "function": {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": getattr(tool, "input_schema", {}) or {},
                },
            }
            for tool in registry.all()
        ]
